README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->
## Features

![image](https://github.com/hhgz9527/yg_indicator_button/blob/main/example.gif?raw=true)

## Getting started

add `yg_indicator_button:` to your `pubspec.yaml` file.

```
dependencies:
  flutter:
    sdk: flutter
  yg_indicator_button:
```

## Usage

TODO: Include short and useful examples for package users. Add longer examples
to `/example` folder. 

```dart
YGIndicatorButton(
    normalTitle: 'Create',
    height: 50,
    width: 200,
    action: (closure) {
        closure(YGIndicatorStatus.loading);
        // fake API request
	Future.delayed(const Duration(seconds: 2), () {
	    closure(YGIndicatorStatus.completed);
	});
    },
)
```


CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* UPDATE: Read me file

## 0.0.3

* ADDED: background color
* ADDED: normalWidget / loadingWidget / completedWidget, when use this, title is disbale


test/yg_indicator_button_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:yg_indicator_button/yg_indicator_button.dart';

void main() {
  test('adds one to input values', () {
  });
}


lib/yg_indicator_button.dart
library yg_indicator_button;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

enum YGIndicatorStatus {
  none,
  loading,
  completed,
}

class YGIndicatorButton extends StatefulWidget {
  final String normalTitle;
  final String? loadingTitle;
  final String? completedTitle;
  final double? width;
  final double? height;
  final Widget? normalWidget; // when use this, normalTitle will disable
  final Widget? loadingWidget; // when use this, loadingTitle will disable
  final Widget? completedWidget; // when use this, completedTitle will disable
  final Color? backgroundColor;
  final Function(Function(YGIndicatorStatus result))? action;

  const YGIndicatorButton({
    Key? key,
    required this.normalTitle,
    this.loadingTitle,
    this.completedTitle,
    this.action,
    this.width,
    this.height,
    this.backgroundColor,
    this.normalWidget,
    this.loadingWidget,
    this.completedWidget,
  }) : super(key: key);

  @override
  _YGIndicatorButtonState createState() => _YGIndicatorButtonState();
}

class _YGIndicatorButtonState extends State<YGIndicatorButton> {
  YGIndicatorStatus _requestStatus = YGIndicatorStatus.none;

  @override
  Widget build(BuildContext context) {
    switch (_requestStatus) {
      case YGIndicatorStatus.none:
        return SizedBox(
          height: widget.height ?? 50,
          width: widget.width ?? double.infinity,
          child: CupertinoButton(
            color: widget.backgroundColor ?? CupertinoColors.systemBlue,
            child: widget.normalWidget ?? Text(widget.normalTitle),
            onPressed: () {
              if (widget.action != null) {
                widget.action!(
                  (requestStatus) {
                    setState(() {
                      _requestStatus = requestStatus;
                    });
                  },
                );
              }
            },
          ),
        );
      case YGIndicatorStatus.loading:
        return Container(
          height: widget.height ?? 50,
          width: widget.width ?? double.infinity,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(8),
            color: widget.backgroundColor ?? CupertinoColors.systemBlue,
          ),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              _buildActivityIndicator(),
              Container(width: 15),
              widget.loadingWidget ?? Text(widget.loadingTitle ?? 'Updating', style: const TextStyle(color: Colors.white)),
            ],
          ),
        );
      case YGIndicatorStatus.completed:
        return Container(
          height: widget.height ?? 50,
          width: widget.width ?? double.infinity,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(8),
            color: Theme.of(context).brightness == Brightness.light ? const Color(0xff23BA66).withAlpha(200) : const Color(0xff23BA66),
          ),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Icon(Icons.done, color: Colors.white),
              Container(width: 15),
              widget.completedWidget ?? Text(widget.completedTitle ?? 'Succeeded', style: const TextStyle(color: Colors.white)),
            ],
          ),
        );
    }
  }

  _buildActivityIndicator() {
    return CupertinoTheme(
      data: CupertinoThemeData(
        brightness: Theme.of(context).brightness,
      ),
      child: const CupertinoActivityIndicator(radius: 8),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:yg_indicator_button/yg_indicator_button.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child: YGIndicatorButton(
          normalTitle: 'Create',
          height: 50,
          width: 200,
          action: (closure) {
            closure(YGIndicatorStatus.loading);
            // fake API request
            Future.delayed(const Duration(seconds: 2), () {
              closure(YGIndicatorStatus.completed);
            });
          },
        ),
      ),
    );
  }
}
